### PR TITLE
Use Dask instead of multiprocessing module

### PR DESCRIPTION
**Pull request by andrii-i**
Use Dask instead of multiprocessing module

### CODE REVIEW

**Comment by andrii-i** [comment]
Kicking CI

File: jupyter_scheduler/extension.py [hunk 1]
```diff
@@ -91,3 +92,25 @@ def initialize_settings(self):
...
+        """
+        Public method called by Jupyter Server when the server is stopping.
+        This calls the cleanup code defined in `self._stop_exception()` inside
+        an exception handler, as the server halts if this method raises an
+        exception.
+        """
+        try:
+            await self._stop_extension()
+        except Exception as e:
+            self.log.error("Jupyter Scheduler raised an exception while stopping:")
```
[the remark this answers is not in the document]
Comment (andrii-i) [reply jupyter_scheduler/extension.py:106]: Thank you for looking into this. Error message in question indicates that Jupyter Scheduler stop_extension logic threw an error. Note that the next line `self.log.exception(e)` would log the actual error. From my point of view this provides an explanation with a sufficient level of details. 

Could you please elaborate on your suggestion? What kind of detail is missing? What kind of message would you suggest?

[the remark this answers is not in the document]
Comment (srdas) [reply jupyter_scheduler/extension.py:106]: Agree, that is all that is needed. Thanks for clarifying.

File: conftest.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 from pathlib import Path
+from unittest.mock import AsyncMock
```
Comment (JasonWeill) [inline conftest.py:2]: This is the only change to this file, and `AsyncMock` isn't used anywhere else in this PR. Should it remain?

File: jupyter_scheduler/scheduler.py [hunk 3]
```diff
@@ -381,6 +382,12 @@ def get_local_output_path(
         else:
             return os.path.join(self.root_dir, self.output_directory, output_dir_name)
 
+    async def stop_extension(self):
+        """
+        Placeholder method for a cleanup code to run when the server is stopping.
+        """
+        pass
```
Comment (JasonWeill) [inline jupyter_scheduler/scheduler.py:389]: Do we need to keep this method if it does nothing?

**Review by JasonWeill — APPROVED** [review]
Generally looking good, just a couple of small non-blocking comments.

File: jupyter_scheduler/scheduler.py (hunk 3 above)
Comment (andrii-i) [reply jupyter_scheduler/scheduler.py:389]: Yes. I [implement](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/scheduler.py#L788-L794) this method in `Scheduler.stop_extension`. Note that `BaseScheduler` is a[ base class / klass for schedulers](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/extension.py#L56-L61), essentially an interface.

To correctly terminate Dask and avoid JupyterLab process not shutting down immediately on interrupt as we've seen in Jupyter AI, I need to [call](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/extension.py#L96-L116) `scheduler.stop_extension` from `SchedulerApp.stop_extension` (a public method inherited from `ExtensionApp` called by Jupyter Server when the server is stopping). Note that `scheduler` in this case is some subclass of `BaseScheduler`.

So `BaseScheduler` base class / klass / interface needs to have a placeholder `stop_extension` method that does nothing by default and that subclasses can implement `stop_extension` logic in.

Comment (andrii-i) [reply jupyter_scheduler/scheduler.py:389]: Yes. `BaseScheduler` is a [base class / klass for schedulers](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/extension.py#L56-L61), essentially an interface for all schedulers [instantiated and used](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/extension.py#L77) by `SchedulerApp`. `BaseScheduler.stop_extension` defines scheduler cleanup / stop_extension logic (if any) and is [called](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/extension.py#L96-L116) from `SchedulerApp.stop_extension`. 
`SchedulerApp.stop_extension` is a public method inherited from `ExtensionApp` called by Jupyter Server when the server is stopping.  

`stop_extension` is [implemented](https://github.com/andrii-i/jupyter-scheduler/blob/4c102d14415aeb6f29b0f1022fddc83c4b0058d6/jupyter_scheduler/scheduler.py#L788-L794)  in default `Scheduler.stop_extension` to correctly terminate Dask and avoid JupyterLab process not shutting down immediately on interrupt as for example we've seen in Jupyter AI.

**Comment by andrii-i** [comment]
Superseded by #553

**Pull request closed without merging**